PWGJE/TableProducer/jettaggerhf.cxx

```diff
@@ -38,48 +42,76 @@ struct JetTaggerHFTask {
       // if (doSV) algorithm2 = jettaggingutilities::Algorithm2((mcdjet, tracks);
-      // if (doML) algorithm3 = jettaggingutilities::Algorithm3((mcdjet, tracks);
-      taggingTableData(0, algorithm1, algorithm2, algorithm3);
+      taggingTableData(0, jetProb, algorithm2, algorithm3);
     }
   }
   PROCESS_SWITCH(JetTaggerHFTask, processData, "Fill tagging decision for data jets", false);
 
-  void processMCD(JetCollision const& collision, JetTableMCD const& mcdjets, JetTracksMCD const& tracks, JetParticles const& particles)
+  void processMCD(JetCollision const& collision, JetTableMCD const& mcdjets, JetTagTracksMCD const& tracks, JetParticles const& particles)
   {
     for (auto& mcdjet : mcdjets) {
-      typename JetTracksMCD::iterator hftrack;
+      typename JetTagTracksMCD::iterator hftrack;
       int origin = 0;
       if (!doWShower)
-        origin = jettaggingutilities::mcdJetFromHFShower(mcdjet, tracks, particles, maxDeltaR); // TODO: it is not working due to getOriginalMotherIndex
+        origin = jettaggingutilities::mcdJetFromHFShower(mcdjet, tracks, particles, maxDeltaR);
       else
         origin = jettaggingutilities::jetTrackFromHFShower(mcdjet, tracks, particles, hftrack);
-      int algorithm1 = 0;
+      float jetProb = 0;
       int algorithm2 = 0;
       int algorithm3 = 0;
-      // if (doTC) algorithm1 = jettaggingutilities::Algorithm1((mcdjet, tracks);
+      TF1* fSignImpXYSig = jettaggingutilities::getResolutionFunction(origin);
+      if (doTC) {
+        jetProb = jettaggingutilities::getJetProbability(fSignImpXYSig, collision, mcdjet, tracks);
+      }
       // if (doSV) algorithm2 = jettaggingutilities::Algorithm2((mcdjet, tracks);
-      // if (doML) algorithm3 = jettaggingutilities::Algorithm3((mcdjet, tracks);
-      taggingTableMCD(origin, algorithm1, algorithm2, algorithm3);
+      taggingTableMCD(origin, jetProb, algorithm2, algorithm3);
     }
   }
   PROCESS_SWITCH(JetTaggerHFTask, processMCD, "Fill tagging decision for mcd jets", false);
+
+  void processTraining(JetCollision const& collision, JetTableMCD const& mcdjets, JetTagTracksMCD const& tracks)
+  {
+    // To create table for ML
+  }
+  PROCESS_SWITCH(JetTaggerHFTask, processTraining, "Fill tagging decision for mcd jets", false);
+};
+
+struct JetTaggerHFExtTask {
+
+  Produces<aod::JTracksTag> jTracksTagTable;
+
+  void init(InitContext const&)
+  {
+  }
+
+  void processTracks(soa::Join<aod::Tracks, aod::TracksCov, aod::TrackSelection, aod::TracksDCA, aod::TracksDCACov>::iterator const& track)
+  {
+    float dcaXYZ = 0;
+    float sigmaDcaXYZ2 = 0;
+
+    jettaggingutilities::calculateDcaXYZ(dcaXYZ, sigmaDcaXYZ2, track.dcaXY(), track.dcaZ(), track.cYY(), track.cZY(), track.cZZ(), track.sigmaDcaXY2(), track.sigmaDcaZ2());
+    jTracksTagTable(track.x(), track.y(), track.z(), track.alpha(), track.snp(), track.tgl(), track.signed1Pt(), track.dcaXY(), track.dcaZ(), dcaXYZ, track.sigmaDcaXY2(), track.sigmaDcaZ2(), sigmaDcaXYZ2);
```

Comment:
Seems you rewrite the information in the tracks table to your own table without calculating something new? why is this done? Cant you just give the original tracks tables to your next task that needs them? Or are the parameters you save different? For example track.alpha() seems the same as the original track one?

Reply:
I know that JTracks can be converted to aod::Tracks at any time
But I thought it would be easier to store the variables I need (dcaXYZ, sigmaDcaXYZ) in a table and join it with JTracks to do the task, so this is how I did it. Is this not an efficient way at all?
I used the table like `using JetTagTracksData = soa::Join<aod::JTracks, aod::JTracksTag>;`

Reply:
Hi Hanseo, yes this is fine for variables that do not exist in the track table like dcaXYZ or sigmaDcaXYZ but for other variables that exist such as x, y, alpha, snp etc i think we should not copy the tables again and should instead get these from aod::Track for now

Reply:
As your comment, I have removed the variables except dcaXYZ and sigmaDcaXYZ2 from the extension table.
However, to do this, in the qa task macro, I used the following: `int geoSign = jettaggingutilities::getGeoSign(collision, jet, track);`
to get
```
for (auto& jtrack : jet.template tracks_as<V>()) {
  auto track = jtrack.template tracks_as<W>();
}
```
I've done the the above in every task.
If there is a way to store the values of geoSign, etc. joinable with JTracks, I would like to do it, but it's not possible within my knowledge because and::JTracks is a all track loop, and variables like geoSign are not the right size because they need to run around the jet loop. If you can suggest a way to do it, I'll go that way. If it's not important, I'll keep jtracks from JTracks and tracks from aod::Tracks separate in the future.
Since I made this change, I also changed the function for jet probability (which also requires a geoSign)
Finally, I also added the starting point for the secondary vertex method (not the IPs method) to the qa macro.
With this pull request, I would like to do a full qa of the IP (track counting, jet probability) method and the SV method.

Reply:
I think for now we should keep it separate like this. In principle i think you should be able to join JTracks and Tracks together anyways (you can maybe try this and see if it works, but it is not garunteed that this will always be true). So for now i think your solution is fine